libevmasm/SemanticInformation.cpp

```diff
@@ -153,6 +153,31 @@ bool SemanticInformation::isDeterministic(AssemblyItem const& _item)
 	}
 }
 
+bool SemanticInformation::movable(Instruction _instruction)
+{
+	// These are not really functional.
+	if (isDupInstruction(_instruction) || isSwapInstruction(_instruction))
+		return false;
+	InstructionInfo info = instructionInfo(_instruction);
+	if (info.sideEffects)
```

Comment:
There are so many of these semi-overlapping definitions now. Is there a way we could simplify all these?

e.g. low-level lists of 
- `readsCallState`
- `readsState`
- `writesState`
- `invalidatesMemory`

and use a combinations of these in these high level helpers.

Reply:
Yes, we probably should. It also turns out that just a boolean `movable` is not that useful. An instruction that reads memory can still be moved as long as no other instruction writes memory in between (or even if we can be sure that the location is different).

A perhaps little different approach might be to just provide a function that can tell whether two opcodes are commutative.

Reply:
Ah and I don't mean commutative in the sense that the arguments of a function can be permuted but in the sense that:
`f` and `g` commute if and only if
```
f(a_1, ..., a_n) g(b_1, ..., b_m)
```
is semantically eqivalent to
```
g(b_1, ..., b_m) f(a_1, ..., a_n) 
```
as long as all of `b_1`, ..., `b_n` commute with all of `a_1`, ..., `a_n`.
Furthermore, literals commute with anything and variables commute with anything as long as they are not re-assigned.

Reply:
And then we can replace `invalidatesMemory` by asking whether it commutes with `mload` (we have to check whether we get problems with msize there) and `invalidatesStorage` is equivalent to commuting with `sload`. Movable in its current form would be "commutes with anything".

I think readState (invalidInPureFunction) and writeState (invalidInViewFunction) still need to be their own function.

Reply:
Created an issue out of this, we can apply those changes separately after this.